File: workers/agent-host/src/connection-configuration.ts
import { readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { join } from 'node:path';
import type { OauthEnv } from '@astra/oauth';

export const CONNECTION_CONFIGURATION_KEYS = new Set([
  'ASTRA_OAUTH_GOOGLE_CLIENT_ID',
  'ASTRA_OAUTH_GOOGLE_CLIENT_SECRET',
  'ASTRA_OAUTH_GOOGLE_READ_CLIENT_ID',
  'ASTRA_OAUTH_GOOGLE_READ_CLIENT_SECRET',
  'ASTRA_OAUTH_GOOGLE_WRITE_CLIENT_ID',
  'ASTRA_OAUTH_GOOGLE_WRITE_CLIENT_SECRET',
  'ASTRA_OAUTH_MICROSOFT_READ_CLIENT_ID',
  'ASTRA_OAUTH_MICROSOFT_WRITE_CLIENT_ID',
]);

/** Native client settings only. User tokens stay in the OS credential store. */
export function connectionConfiguration(
  env: OauthEnv,
  platform = process.platform,
  home = homedir(),
): OauthEnv {
  const path =
    env['ASTRA_CONNECTIONS_CONFIG'] ??
    (platform === 'darwin' && !env['ASTRA_SECRET_STORE_FILE'] && !env['ASTRA_DATA_ROOT']
      ? join(home, 'Library/Application Support/Astra/connections.json')
      : undefined);
  let local: Record<string, string> = {};
  if (path) {
    try {
      const value: unknown = JSON.parse(readFileSync(path, 'utf8'));
      if (value && typeof value === 'object' && !Array.isArray(value)) {
        local = Object.fromEntries(
          Object.entries(value)
            .filter(
              ([key, value]) => CONNECTION_CONFIGURATION_KEYS.has(key) && typeof value === 'string',
            )
            .map(([key, value]) => [key, (value as string).trim()]),
        );
      }
    } catch {
      /* Missing/invalid configuration stays unavailable; never log credentials. */
    }
  }
  return {
    ...local,
    ...Object.fromEntries(Object.entries(env).filter(([, value]) => value !== undefined)),
  };
}

File: apps/astra-macos/Sources/AstraMac/Context/ConnectionConfiguration.swift
import Foundation

/// Publisher-owned native OAuth configuration. Never contains user tokens.
/// The same allowlisted JSON file is read by the local agent host for refresh.
enum ConnectionConfiguration {
    static let keys: Set<String> = [
        "ASTRA_OAUTH_GOOGLE_CLIENT_ID", "ASTRA_OAUTH_GOOGLE_CLIENT_SECRET",
        "ASTRA_OAUTH_GOOGLE_READ_CLIENT_ID", "ASTRA_OAUTH_GOOGLE_READ_CLIENT_SECRET",
        "ASTRA_OAUTH_GOOGLE_WRITE_CLIENT_ID", "ASTRA_OAUTH_GOOGLE_WRITE_CLIENT_SECRET",
        "ASTRA_OAUTH_MICROSOFT_READ_CLIENT_ID", "ASTRA_OAUTH_MICROSOFT_WRITE_CLIENT_ID",
    ]
    static var localURL: URL {
        FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Astra/connections.json")
    }
    static func merge(_ layers: [[String: String]]) -> [String: String] {
        var result: [String: String] = [:]
        for layer in layers {
            for (key, value) in layer where keys.contains(key) {
                result[key] = value.trimmingCharacters(in: .whitespacesAndNewlines)
            }
        }
        return result
    }
    static func read(_ url: URL?) -> [String: String] {
        guard let url, let data = try? Data(contentsOf: url),
              let values = try? JSONDecoder().decode([String: String].self, from: data) else { return [:] }
        return merge([values])
    }
    static var current: [String: String] {
        let env = ProcessInfo.processInfo.environment
        // Fixtures must not consume the user's real configuration.
        let isolated = env["ASTRA_DATA_ROOT"] != nil || CommandLine.arguments.contains("--selftest")
        let explicit = env["ASTRA_CONNECTIONS_CONFIG"].map { URL(fileURLWithPath: $0) }
        return merge([
            isolated ? [:] : read(Bundle.main.url(forResource: "connections", withExtension: "json")),
            read(explicit ?? (isolated ? nil : localURL)), env,
        ])
    }
    static func clientId(provider: String, readOnly: Bool, values: [String: String]) -> String? {
        let prefix = "ASTRA_OAUTH_" + provider.uppercased()
        let read = values[prefix + "_READ_CLIENT_ID"] ?? (provider == "google" ? values[prefix + "_CLIENT_ID"] : nil)
        let write = values[prefix + "_WRITE_CLIENT_ID"] ?? (provider == "google" ? values[prefix + "_CLIENT_ID"] : nil)
        if provider == "microsoft", let read, !read.isEmpty, read == write { return nil }
        return (readOnly ? read : write).flatMap { $0.isEmpty ? nil : $0 }
    }
    /// The agent host also needs the publisher's parameters to renew a saved grant.
    static func persistForRefresh() throws {
        let values = current
        guard !values.isEmpty else { return }
        try FileManager.default.createDirectory(at: localURL.deletingLastPathComponent(), withIntermediateDirectories: true)
        try JSONEncoder().encode(values).write(to: localURL, options: .atomic)
        try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: localURL.path)
    }
    static func clientSecret(provider: String, readOnly: Bool, values: [String: String]) -> String? {
        guard provider == "google" else { return nil }
        let role = readOnly ? "READ" : "WRITE"
        if values["ASTRA_OAUTH_GOOGLE_\(role)_CLIENT_ID"] != nil {
            return values["ASTRA_OAUTH_GOOGLE_\(role)_CLIENT_SECRET"]
        }
        return values["ASTRA_OAUTH_GOOGLE_CLIENT_SECRET"]
    }
}
